drivers/DeviceKey.cpp

```diff
@@ -0,0 +1,255 @@
+{
+    if (size > UINT16_MAX) {
+        return DEVICEKEY_INVALID_PARAM;
+    }
+
+    uint16_t in_size = size;
+    uint16_t out_size = 0;
+    NVStore& nvstore = NVStore::get_instance();
+    int nvStatus = nvstore.get(NVSTORE_DEVICEKEY_KEY, in_size, output, out_size);
+    if (NVSTORE_NOT_FOUND == nvStatus) {
+        return DEVICEKEY_NOT_FOUND;
+    }
+
+    if (NVSTORE_READ_ERROR == nvStatus || NVSTORE_BUFF_TOO_SMALL == nvStatus) {
+        return DEVICEKEY_READ_FAILED;
+    }
+
+    if (NVSTORE_SUCCESS != nvStatus) {
+        return DEVICEKEY_NVSTORE_UNPREDICTABLE_ERROR;
+    }
+
+    size = out_size;
+    return DEVICEKEY_SUCCESS;
+}
+
+// Calculate CMAC functions - wrapper for mbedtls start/update and finish
+int DeviceKey::calculate_cmac(const unsigned char *input, size_t isize, uint32_t *ikey_buff, int ikey_size,
+                              unsigned char *output)
+{
+    int ret;
+    mbedtls_cipher_context_t ctx;
+
+    mbedtls_cipher_type_t mbedtls_cipher_type = MBEDTLS_CIPHER_AES_128_ECB;
+    if (DEVICE_KEY_32BYTE == ikey_size) {
+        mbedtls_cipher_type = MBEDTLS_CIPHER_AES_256_ECB;
+    }
+
+    const mbedtls_cipher_info_t *cipher_info = mbedtls_cipher_info_from_type(mbedtls_cipher_type);
+
+    mbedtls_cipher_init(&ctx);
+    ret = mbedtls_cipher_setup(&ctx, cipher_info);
+    if (ret != 0) {
+        goto finish;
+    }
+
+    ret = mbedtls_cipher_cmac_starts(&ctx, (unsigned char *)ikey_buff, ikey_size * 8);
+    if (ret != 0) {
+        goto finish;
+    }
+
+    ret = mbedtls_cipher_cmac_update(&ctx, input, isize);
+    if (ret != 0) {
+        goto finish;
+    }
+
+    ret = mbedtls_cipher_cmac_finish(&ctx, output);
+    if (ret != 0) {
+        goto finish;
+    }
+
+    return DEVICEKEY_SUCCESS;
+
+finish:
+    mbedtls_cipher_free( &ctx );
+    return ret;
+}
+
+int DeviceKey::get_derived_key(uint32_t *ikey_buff, size_t ikey_size, const unsigned char *isalt,
+                               size_t isalt_size, unsigned char *output, uint32_t ikey_type)
+{
+    int ret;
+    unsigned char *double_size_salt = NULL;
+
+    if (DEVICE_KEY_16BYTE == ikey_type) {
+        ret = calculate_cmac(isalt, isalt_size, ikey_buff, ikey_size, output);
+        if (DEVICEKEY_SUCCESS != ret) {
+            goto finish;
+        }
+    }
+
+    if (DEVICE_KEY_32BYTE == ikey_type) {
+        ret = this->calculate_cmac(isalt, isalt_size, ikey_buff, ikey_size, output);
+        if (DEVICEKEY_SUCCESS != ret) {
+            goto finish;
+        }
+
+        //Double the salt size cause cmac always return just 16 bytes
+        double_size_salt = new unsigned char[isalt_size * 2];
+        memcpy(double_size_salt, isalt, isalt_size);
+        memcpy(double_size_salt + isalt_size, isalt, isalt_size);
+
+        ret = this->calculate_cmac(double_size_salt, isalt_size * 2, ikey_buff, ikey_size, output + 16);
+    }
+
+finish:
+    if (double_size_salt != NULL) {
+        delete[] double_size_salt;
```

Comment:
Please reliably zeroize `double_size_salt` to avoid leakage in memory. Consider e.g. `mbedtls_zeroize`.

Reply:
@hanno-arm 
mbedtls_zeroize is static function duplicate inside every mbedtls module. For now I have copied it as static function to my code. I think we should discuss on where this function should sit in order to have only one copy for everyone.

Reply:
@yossi2le Moving the zeroize function to a separate compilation unit is being considered at the moment. For now, I suggest we live with the additional copy here.